src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,466 @@
+                DataType::Boolean,
+                Box::new(InputRefExpression::new(
+                    time_col_data_type.clone(),
+                    self.time_col_idx,
+                )),
+                Box::new(LiteralExpression::new(
+                    DataType::Timestamp,
+                    Some(ScalarImpl::NaiveDateTime(timestamp_interval_sub::<
+                        NaiveDateTimeWrapper,
+                        IntervalUnit,
+                        NaiveDateTimeWrapper,
+                    >(
+                        NaiveDateTimeWrapper::new(NaiveDateTime::from_timestamp(
+                            (time_millis / 1000) as i64,
+                            (time_millis % 1000 * 1_000_000) as u32,
+                        )),
+                        window_size,
+                    )?)),
+                )),
+            )
+        };
+
+        let mut input = self.input.take().unwrap().execute();
+
+        // Consume the first barrier message and initialize state table.
+        let barrier = expect_first_barrier(&mut input).await?;
+        self.state_table.init_epoch(barrier.epoch);
+        let mut time_millis = Epoch::from(barrier.epoch.curr).as_unix_millis();
+        let mut sync_expr = None;
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
+
+        // Fill the buffer on initialization. If the executor has recovered from a failed state, the
+        // buffer should be refilled by its previous data.
+        let vnode_bitmap = self.state_table.vnode_bitmap().to_owned();
+        self.fill_buffer(None, &vnode_bitmap).await?;
+
+        #[for_await]
+        for msg in input {
+            match msg? {
+                Message::Watermark(_) => {
+                    todo!("https://github.com/risingwavelabs/risingwave/issues/6042")
+                }
+                Message::Chunk(chunk) => {
+                    if sync_expr.is_none() {
+                        sync_expr = Some(dynamic_cond(time_millis)?);
+                        for stream_chunk in self.sync(sync_expr.as_ref().unwrap())? {
+                            yield Message::Chunk(stream_chunk);
+                        }
+                    }
+
+                    let (data_chunk, ops) = chunk.into_parts();
+                    let (new_ops, new_visibility) =
+                        self.apply_batch(&data_chunk, ops, sync_expr.as_ref().unwrap());
+                    if new_visibility.num_high_bits() > 0 {
+                        let (columns, _) = data_chunk.into_parts();
+                        let new_chunk = StreamChunk::new(new_ops, columns, Some(new_visibility));
+                        yield Message::Chunk(new_chunk);
+                    }
+                }
+                Message::Barrier(barrier) => {
+                    time_millis = Epoch::from(barrier.epoch.curr).as_unix_millis();
+                    if barrier.is_update() {
+                        sync_expr = None;
```

Comment:
Why do we not simply perform the sync regardless of update or not, after filling/cleaning the buffer/cache upon update?

Reply:
In fact, I think it doesn't even matter whether we perform the sync before or after the update. The result should be the same.

Reply:
> Why do we not simply perform the sync regardless of update or not, after filling/cleaning the buffer/cache upon update?
> 
> Then, we can simply remove the `if sync_expr.is_none()`. We can also always assume `sync_expr` is Some, i.e. it does not need to be an option?

This is deliberate because when mutation occurs, it will perform 'pause barrier - update barrier - resume barrier' stages. We are not permitted to emit any `Chunk` throughout these stages.

Reply:
I see, wouldn't we simply sync on resume barrier?